### PR TITLE
Only logged in users

### PR DESCRIPTION
**Pull request by Mentessi**
Changes relevant to this pull request:
https://github.com/Mentessi/toastmasters/compare/user-authentication...only-logged-in-users

### CODE REVIEW

File: app/views/welcome/index.html.erb [hunk 1]
```diff
@@ -1,4 +1,16 @@
...
+</p>
+
+<%= link_to 'View Topics', topics_url %>
+</br>
+<% if user_signed_in? %>
+  <%= link_to 'Log Out', destroy_user_session_url, method: :delete %>
+  <% else %>
+  <%= link_to 'Sign Up', new_user_registration_url %>
+  </br>
+  <%= link_to 'Log In', new_user_session_url %>
```
Comment (Neilos) [inline app/views/welcome/index.html.erb]: I think the 'Log In' link should only show when the user is not signed in, so you need an else statement.

File: app/views/welcome/index.html.erb [hunk 2]
```diff
@@ -1,4 +1,16 @@
...
   The time is now: <%= Time.now %>
-</p>
\ No newline at end of file
+</p>
+
+<%= link_to 'View Topics', topics_url %>
+</br>
+<% if user_signed_in? %>
+  <%= link_to 'Log Out', destroy_user_session_url, method: :delete %>
+  <% else %>
```
Comment (Neilos) [inline app/views/welcome/index.html.erb]: Outdent this `else`

File: spec/requests/requests_spec.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
+
+RSpec.describe 'managing topics', :type => :request do
+
+  let!(:user) {FactoryBot.create :user}
+
+  it 'redirects unauthenticated users' do
+    post "/topics", :params => { :topic => {:name => "here is a new topic"} }
+    expect(response).to redirect_to('/users/sign_in')
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: You could use the route helper if you want (`new_session_path` I think)

File: spec/requests/requests_spec.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  it 'allows topic posting for logged in users' do
+    post user_session_path, :params => {'user[email]' => user.email, 'user[password]' => user.password }
+    get '/topics/new'
+    expect(response).to render_template(:new)
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: I'd just expect the response *not* to be a redirect to the new_session_path. The internals of where users get sent/redirected to when they are properly authenticated and authorised is covered by feature/system specs.

File: spec/requests/requests_spec.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
+
+RSpec.describe 'managing topics', :type => :request do
+
+  let!(:user) {FactoryBot.create :user}
+
+  it 'redirects unauthenticated users' do
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: Could do with a describe block to say what we're testing (i.e. post#create)

File: spec/requests/requests_spec.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    post "/topics", :params => { :topic => {:name => "here is a new topic"} }
+    expect(response).to redirect_to('/users/sign_in')
+  end
+
+  it 'allows topic posting for logged in users' do
+    post user_session_path, :params => {'user[email]' => user.email, 'user[password]' => user.password }
+    get '/topics/new'
+    expect(response).to render_template(:new)
+
+    post "/topics", :params => { :topic => {:name => "here is a new topic"} }
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: Don't test multiple requests in one spec (e.g in this test your testing a get#new and a post#create request).
For the test structure I would do a describe for each controller action.

```ruby
describe 'put#update' do
  it 'is forbidden for logged out users' do
    ...
  end
  
  it 'is allowed for authenticated users' do
    ...
  end
end

File: spec/requests/requests_spec.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    expect(response).to redirect_to('/topics')
+    follow_redirect!
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: Don't need to follow_redirects if you just test the response and only test one request per spec

File: spec/requests/requests_spec.rb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  let!(:user) {FactoryBot.create :user}
+
+  it 'redirects unauthenticated users' do
+    post "/topics", :params => { :topic => {:name => "here is a new topic"} }
+    expect(response).to redirect_to('/users/sign_in')
+  end
+
+  it 'allows topic posting for logged in users' do
+    post user_session_path, :params => {'user[email]' => user.email, 'user[password]' => user.password }
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: Don't do user authentication like this (unless it is the thing you are testing - which it isn't here).

Instead use warden test helpers provided by Devise. See [this](https://github.com/plataformatec/devise/wiki/How-To:-Test-with-Capybara) for some clues - it's talking about capybara, which isn't relevant, but the helpers should work for request specs.

File: spec/requests/requests_spec.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    get '/topics/new'
+    expect(response).to render_template(:new)
+
+    post "/topics", :params => { :topic => {:name => "here is a new topic"} }
+    expect(response).to redirect_to('/topics')
+    follow_redirect!
+
+    expect(response.body).to include("here is a new topic")
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: Don't need this assertion. We just care about whether the response was a redirect to login.

File: Gemfile [hunk 10]
```diff
@@ -31,6 +31,7 @@ gem 'jbuilder', '~> 2.5'
 # Use ActiveModel has_secure_password
 # gem 'bcrypt', '~> 3.1.7'
 gem 'devise'
+gem 'rails-controller-testing'
```
Comment (Neilos) [inline Gemfile]: You shouldn't use this gem anymore. You don't need it for request specs. 

It is a gem to restore some functionality that was removed from rails.
Basically, rails has removed controller specs, which used to reach into the controller class and check the value of instance variables (`with assigns`) - very weird hence its removal. This gem restores that weird ability 👎 .

File: spec/requests/requests_spec.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
+require 'rails_helper'
+
+RSpec.describe 'managing topics', :type => :request do
```
Comment (Neilos) [inline spec/requests/requests_spec.rb]: I generally name my request specs after the controller -  Topic requests - One set of request specs per controller.
Also, name your request spec file that way (e.g. _topic_requests_spec.rb_)

File: Gemfile (hunk 10 above)
Comment (Mentessi) [reply Gemfile]: I installed the gem as I get this error without it - 
```
Failure/Error: expect(response).to render_template(:new)

NoMethodError:
       assert_template has been extracted to a gem. To continue using it,
               add `gem 'rails-controller-testing'` to your Gemfile.
```

Comment (Mentessi) [reply Gemfile]: I have just removed the assert_template related assertions from my tests :)

File: spec/requests/requests_spec.rb (hunk 3 above)
Comment (Mentessi) [reply spec/requests/requests_spec.rb]: new_user_session_path werks ;)

File: Gemfile (hunk 10 above)
Comment (Neilos) [reply Gemfile]: `assert_template` was another weird rails method that allowed you to check what template the controller renders.

The reason this (and `assigns`) are weird is that they are checking not the return value of the method or the side effects of the method, but instead are checking the bits of logic *inside* the method, which normal unit tests and feature testing never do.

**Pull request merged**